lib/meilisearch/http_request.rb

```diff
@@ -7,109 +7,112 @@ module MeiliSearch
     end
 
     def http_get(relative_path = '', query_params = {})
       send_request(
         proc { |path, config| self.class.get(path, config) },
         relative_path,
-        query_params: query_params,
-        options: remove_options_header(@options, 'Content-Type')
+        config: {
+          query_params: query_params,
+          headers: remove_headers(@headers.dup, 'Content-Type'),
+          options: @options
+        }
       )
     end
 
     def http_post(relative_path = '', body = nil, query_params = nil, options = {})
       send_request(
         proc { |path, config| self.class.post(path, config) },
         relative_path,
-        query_params: query_params,
-        body: body,
-        options: merge_options(@options, options)
+        config: {
+          query_params: query_params,
+          body: body,
+          headers: @headers.dup.merge(options[:headers] || {}),
+          options: @options.merge(options)
+        }
       )
     end
 
     def http_put(relative_path = '', body = nil, query_params = nil)
       send_request(
         proc { |path, config| self.class.put(path, config) },
         relative_path,
-        query_params: query_params,
-        body: body,
-        options: @options
+        config: {
+          query_params: query_params,
+          body: body,
+          headers: @headers,
+          options: @options
+        }
       )
     end
 
     def http_patch(relative_path = '', body = nil, query_params = nil)
       send_request(
         proc { |path, config| self.class.patch(path, config) },
         relative_path,
-        query_params: query_params,
-        body: body,
-        options: @options
+        config: {
+          query_params: query_params,
+          body: body,
+          headers: @headers,
+          options: @options
+        }
       )
     end
 
     def http_delete(relative_path = '')
       send_request(
         proc { |path, config| self.class.delete(path, config) },
         relative_path,
-        options: remove_options_header(@options, 'Content-Type')
+        config: {
+          headers: remove_headers(@headers.dup, 'Content-Type'),
+          options: @options
+        }
       )
     end
 
     private
 
-    def build_default_options_headers(api_key = nil)
-      header = {
-        'Content-Type' => 'application/json'
-      }
-      header = header.merge('Authorization' => "Bearer #{api_key}") unless api_key.nil?
-      header
-    end
-
-    def merge_options(default_options, added_options = {})
-      default_cloned_headers = default_options[:headers].clone
-      merged_options = default_options.merge(added_options)
-      merged_options[:headers] = default_cloned_headers.merge(added_options[:headers]) if added_options.key?(:headers)
-      merged_options
+    def build_default_options_headers
+      {
+        'Content-Type' => 'application/json',
+        'Authorization' => ("Bearer #{@api_key}" unless @api_key.nil?)
```

Comment:
That I was trying to show to you @curquiza that day [#273 comment](https://github.com/meilisearch/meilisearch-ruby/pull/273/files#r779085520)

Reply:
Sooooooo cool!